### ISSUE TITLE
Full restructure

### ISSUE THREAD

**Issue by aumetra** [kitsune-soc/kitsune#418]
So, Kitsune right now is pretty intertwined with ActivityPub and Webfinger. Kinda makes sense since it started out by being built around those standards, but I think we can do better with minimal performance impact.

## Rough plan

1. Move ActivityPub logic into `kitsune-activitypub`
2. Define traits in `kitsune-core`, such as the `Federator`, `FederationFetcher`, and `Resolver` trait
    - These three areas were chosen since they define pretty well how ActivityPub works:
        1. Resolve an actor via Webfinger
        2. Fetch them via a fetcher
        3. Use the federator to send out things such as post creation, deletion, favourites, etc.
3. Instead of giving services a `Fetcher` or `Deliverer`, we give them a vector of these which they iterate over and perform the appropriate actions in the different federated networks
4. Restructure the database
    1. Create a separate key table to store the keys and their IDs and references to an account that owns them
    2. Move the ActivityPub-specific things, such as collection URLs into its seperate table
    3. Rename the `actor_type` field to something more generic (such as just `type`)
    4. Add a `protocol` marker to each account, post, like, etc. to identify from what network they originate from

---

I invite some bikeshedding to this issue since this is a *large* undertaking and will make *significant* changes to all parts of Kitsune

**Comment by aumetra** [comment]
This we still need to address here are system actors (cc @perillamint, if you have any suggestions)

**Comment by aumetra** [comment]
For multiple domains, we could have a table such as this one:

```sql
CREATE TABLE domains (
    domain TEXT PRIMARY KEY,
    owner_id UUID REFERENCES users(id),
    globally_usable BOOLEAN NOT NULL, -- Name can be bikeshed. Marker on whether anyone from the instance can choose this as "their" domain
    verification_token TEXT NOT NULL,
    verified_at TIMESTAMPTZ
);

**Comment by aumetra** [comment]
(gonna try to land #405 before starting this because rebasing on these planned changes would be a *pain*)

**Comment by Zeerooth** [comment]
If we're doing this then I think it's important to think about how to potentially tie https://github.com/kitsune-soc/kitsune/issues/134 into our future account schema

**Comment by aumetra** [comment]
> it's important to think about how to potentially tie https://github.com/kitsune-soc/kitsune/issues/134 into our future account schema

Do you have any suggestions on how we can tie this in?  
Potentially a new table, such as `account_proofs` with the structure of:

```sql
CREATE TABLE accounts_proofs (
    id UUID PRIMARY KEY,
    account_id UUID NOT NULL REFERENCES accounts(id),
    type TEXT NOT NULL,
    verification_method TEXT NOT NULL,
    value TEXT NOT NULL,
    purpose TEXT NOT NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP
);
```

That way we can have a 1:N relationship with accounts, allowing an account to display N proofs

**Comment by aumetra** [comment]
In general, I'd only store the "core" functionality inside the `accounts` table from now on.  
Things that should go into a different table are:

- Collection URLs (followers, following, inbox, outbox, etc.)
- Cryptographic keys
- Even (potentially) the URL of the actor, since we maybe want to serve the users from different URLs dependent on the protocol

By having that neat separation, we don't need to store everything into a single gigantic table

**Comment by silverpill** [comment]
>That way we can have a 1:N relationship with accounts, allowing an account to display N proofs

The relationship between actors and proof subjects (DIDs) is "many to many". Each actor can have multiple proofs, and different actors can be linked to a single subject (this enables migration procedures like the one described in [FEP-7628](https://codeberg.org/fediverse/fep/src/branch/main/fep/7628/fep-7628.md) (sub-type 2)).

ActivityPub 2.0 may look like that: https://codeberg.org/fediverse/fep/src/branch/main/fep/ae97/fep-ae97.md#server-independent-ids (DIDs instead of URLs).

I haven't yet decided how to represent all these possibilities in the database... Moving non-core functionality into separate tables sounds like a good idea.

**Comment by aumetra** [comment]
> The relationship between actors and proof subjects (DIDs) is "many to many"

Yeah, I focussed way too much on the `proof` object. Bue yeah, the `identity` object itself should obviously be serialized to the database, too.

Something like:

```sql
CREATE TABLE accounts_identities (
    id TEXT NOT NULL,
    account_id UUID NOT NULL REFERENCES accounts(id),
    proof_id UUID NOT NULL REFERENCES accounts_proofs(id),
    created_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP,

    PRIMARY KEY (id, account_id)
);
```

That way we can map accounts to identities in an M:N way, and attach a link to a proof as metadata.

**Comment by aumetra** [comment]
Technically this design would be a logical 1:1 relationship between an identity and proof, while you could *technically* generate multiple proofs of the same identity.  
Identity management is a pain.

**Comment by perillamint** [comment]
> We still need to address here are system actors (cc @perillamint, if you have any suggestions). Another unaddressed area is support for multiple domains.
> 
> Edit: Regarding the keys for the system actor, the design outlined [here](https://github.com/kitsune-soc/kitsune/issues/415#issuecomment-1806398558) somewhat incorporates the idea of keys that aren't owned by an actual "account"

I guess system actors could be implemented using some kind of roles table for the accounts, like Mastodon's `user_roles` table.

**Comment by erlend-sh** [comment]
This doc gives a good idea of alternative protocol requirements:
- https://github.com/bluesky-social/atproto/discussions/2350
